# TO DO

# Импорты

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import OrdinalEncoder, MinMaxScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn.impute import KNNImputer

In [2]:
coordinates = [
    {'Location': 'Penrith', 'Latitude': -33.75, 'Longitude': 150.7, 'Elevation': 25},
    {'Location': 'Dartmoor', 'Latitude': -37.92, 'Longitude': 141.27, 'Elevation': 69},
    {'Location': 'Moree', 'Latitude': -29.46, 'Longitude': 149.85, 'Elevation': 212},
    {'Location': 'Bendigo', 'Latitude': -36.76, 'Longitude': 144.28, 'Elevation': 213},
    {'Location': 'Watsonia', 'Latitude': -37.71, 'Longitude': 145.08, 'Elevation': 63},
    {'Location': 'NorahHead', 'Latitude': -33.28, 'Longitude': 151.57, 'Elevation': 20},
    {'Location': 'Hobart', 'Latitude': -42.88, 'Longitude': 147.32, 'Elevation': 51},
    {'Location': 'Albany', 'Latitude': -35.02, 'Longitude': 117.88, 'Elevation': 77},
    {'Location': 'BadgerysCreek', 'Latitude': -33.87, 'Longitude': 150.77, 'Elevation': 50},
    {'Location': 'Ballarat', 'Latitude': -37.56, 'Longitude': 143.85, 'Elevation': 435},
    {'Location': 'SydneyAirport', 'Latitude': -33.94, 'Longitude': 151.18, 'Elevation': 6},
    {'Location': 'Launceston', 'Latitude': -41.43, 'Longitude': 147.14, 'Elevation': 15},
    {'Location': 'Portland', 'Latitude': -38.34, 'Longitude': 141.6, 'Elevation': 32},
    {'Location': 'Wollongong', 'Latitude': -34.42, 'Longitude': 150.89, 'Elevation': 10},
    {'Location': 'Newcastle', 'Latitude': -32.93, 'Longitude': 151.78, 'Elevation': 9},
    {'Location': 'NorfolkIsland', 'Latitude': -29.03, 'Longitude': 167.94, 'Elevation': 110},
    {'Location': 'Brisbane', 'Latitude': -27.47, 'Longitude': 153.03, 'Elevation': 27},
    {'Location': 'MelbourneAirport', 'Latitude': -37.67, 'Longitude': 144.83, 'Elevation': 132},
    {'Location': 'GoldCoast', 'Latitude': -28.02, 'Longitude': 153.4, 'Elevation': 5},
    {'Location': 'MountGinini', 'Latitude': -35.53, 'Longitude': 148.77, 'Elevation': 1760},
    {'Location': 'AliceSprings', 'Latitude': -23.7, 'Longitude': 133.88, 'Elevation': 545},
    {'Location': 'Woomera', 'Latitude': -31.1, 'Longitude': 136.8, 'Elevation': 167},
    {'Location': 'PearceRAAF', 'Latitude': -31.67, 'Longitude': 116.02, 'Elevation': 49},
    {'Location': 'Katherine', 'Latitude': -14.47, 'Longitude': 132.25, 'Elevation': 108},
    {'Location': 'Adelaide', 'Latitude': -34.93, 'Longitude': 138.6, 'Elevation': 48},
    {'Location': 'Darwin', 'Latitude': -12.45, 'Longitude': 130.83, 'Elevation': 30},
    {'Location': 'Melbourne', 'Latitude': -37.81, 'Longitude': 144.96, 'Elevation': 31},
    {'Location': 'Williamtown', 'Latitude': -32.8, 'Longitude': 151.83, 'Elevation': 9},
    {'Location': 'Mildura', 'Latitude': -34.23, 'Longitude': 142.08, 'Elevation': 53},
    {'Location': 'Sale', 'Latitude': -38.1, 'Longitude': 147.07, 'Elevation': 5},
    {'Location': 'Albury', 'Latitude': -36.08, 'Longitude': 146.92, 'Elevation': 165},
    {'Location': 'Cairns', 'Latitude': -16.92, 'Longitude': 145.78, 'Elevation': 7},
    {'Location': 'WaggaWagga', 'Latitude': -35.12, 'Longitude': 147.37, 'Elevation': 212},
    {'Location': 'Witchcliffe', 'Latitude': -34.03, 'Longitude': 115.1, 'Elevation': 68},
    {'Location': 'Tuggeranong', 'Latitude': -35.42, 'Longitude': 149.07, 'Elevation': 578},
    {'Location': 'CoffsHarbour', 'Latitude': -30.3, 'Longitude': 153.12, 'Elevation': 21},
    {'Location': 'PerthAirport', 'Latitude': -31.93, 'Longitude': 115.97, 'Elevation': 20},
    {'Location': 'SalmonGums', 'Latitude': -32.98, 'Longitude': 121.65, 'Elevation': 307},
    {'Location': 'Sydney', 'Latitude': -33.87, 'Longitude': 151.21, 'Elevation': 58},
    {'Location': 'MountGambier', 'Latitude': -37.83, 'Longitude': 140.78, 'Elevation': 63},
    {'Location': 'Cobar', 'Latitude': -31.5, 'Longitude': 145.83, 'Elevation': 260},
    {'Location': 'Canberra', 'Latitude': -35.28, 'Longitude': 149.13, 'Elevation': 577},
    {'Location': 'Nuriootpa', 'Latitude': -34.47, 'Longitude': 139.0, 'Elevation': 274},
    {'Location': 'Townsville', 'Latitude': -19.25, 'Longitude': 146.8, 'Elevation': 15},
    {'Location': 'Perth', 'Latitude': -31.95, 'Longitude': 115.86, 'Elevation': 31},
    {'Location': 'Richmond', 'Latitude': -33.6, 'Longitude': 150.75, 'Elevation': 25},
    {'Location': 'Walpole', 'Latitude': -34.98, 'Longitude': 116.73, 'Elevation': 68},
    {'Location': 'Nhil', 'Latitude': -36.33, 'Longitude': 141.65, 'Elevation': 77},
    {'Location': 'Uluru', 'Latitude': -25.34, 'Longitude': 131.04, 'Elevation': 863},
]

# EDA

Скор на кросс-валидации без обработки данных(заполнение пропусков: -100000) = 0.85422213, 0.85497564, 0.85412167, 0.8543226, 0.8538705
Классы не сбалансированы: 'YES' = 22%; 'NO'= 78%
Заполнение медианой не сильно увеличило скор модели, по сравнению с заполнением большим отрицательным значением

In [3]:
df = pd.read_csv('datasets/train_df.csv')
coordinate_df = pd.DataFrame(coordinates)

In [4]:
df.head()

Date  Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2014-05-24   Penrith     10.6     24.9       0.0          NaN       NaN   
1  2016-09-04  Dartmoor      6.0     15.1       1.6          NaN       NaN   
2  2011-01-26     Moree     24.6     41.4       0.0         12.4      12.8   
3  2013-08-22   Bendigo      3.7     12.6       6.6          1.4       NaN   
4  2011-10-19  Watsonia      7.6     27.9       0.0          8.0      12.5   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0          NW           20.0        SSW  ...        68.0         47.0   
1          NW           41.0        NNW  ...        94.0         77.0   
2          NE           31.0        NNE  ...        51.0         24.0   
3          NW           41.0        WNW  ...        95.0         71.0   
4           N           54.0         NE  ...        57.0         36.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0          NaN          NaN       NaN       NaN     16.7     23.3         No   
1       1023.0       1024.0       NaN       NaN     10.6     13.9        Yes   
2       1012.7       1009.6       0.0       1.0     30.1     39.3         No   
3       1011.3       1009.4       8.0       8.0      8.5     12.2        Yes   
4       1026.6       1023.1       1.0       1.0     19.8     27.3         No   

   RainTomorrow  
0            No  
1            No  
2            No  
3           Yes  
4            No  

[5 rows x 23 columns]

In [5]:
#Соединяем два датафрейма
df = pd.merge(df, coordinate_df, how='outer', on='Location')

In [6]:
#df['RainTomorrow'].plot(kind='hist')
df['Date'] = pd.to_datetime(df['Date'], format='%Y-%m-%d')

In [7]:
category_list = ['Location', 'RainToday', 'RainTomorrow', 'WindGustDir', 'WindDir9am', 'WindDir3pm']

In [8]:
#Отмечаем, было ли значние NaN
for column in df.columns:
    if df[column].isna().sum() > 0:
        df['was_NA_' + column] = 0
        df.loc[df[column].isna() == True, 'was_NA_' + column] = 1

# Заполнение пропусков простой вставкой занчения

In [9]:
comment = '''
#Категориальные заполняем Unknown, а числовые - медианой
df = df.apply(lambda x: x.fillna('Unknown') if x.name in category_list else x.fillna(x.median()))
'''

# Encoding

In [10]:
encoder = OrdinalEncoder(dtype=np.float64)
df[category_list] = encoder.fit_transform(df[category_list])

df = df.rename(columns={'Location' : 'loc_id'})

#Так получается, что столбцу RainToday OrdinalEncoder присваивает в случае No = 0, Yes = 2, Unknown = 1. Меняем значение Unknown на -1, а Yes на 1
df.loc[df.RainToday == 1, 'RainToday'] = -1
df.loc[df.RainToday == 2, 'RainToday'] = 1

In [11]:
df['year'] = df['Date'].apply(lambda x: x.year)
df['month'] = df['Date'].apply(lambda x: x.month)
df['day_of_week'] = df['Date'].apply(lambda x: x.dayofweek)
df['week_of_year'] = df['Date'].apply(lambda x: x.weekofyear)
df = df.sort_values(['loc_id', 'Date']).reset_index().drop('index', axis=1)

# Заполнение пропусков методом KNN

In [12]:
without_date = df.columns.values[1:] #Необходимо для работы скейлера и KNN
num_features = df.select_dtypes(include=['int64', 'float64']).columns.values

scaler = MinMaxScaler(feature_range=(0, 1))
df[num_features] = scaler.fit_transform(df[num_features])

In [13]:
imputer=KNNImputer(n_neighbors=2, weights='uniform')
imputer.fit(df[without_date])

df[without_date] = pd.DataFrame(imputer.transform(df[without_date]), index=df[without_date].index, columns = df[without_date].columns)

# Поиск разницы между двумя ближайшими записями

In [14]:
#Разница между двумя ближайшими записями в датасете для каждого города

#Создаем список с датами для каждого loc_id
gap_days = df.groupby('loc_id').Date.apply(list)

#Дублируем для каждого loc_id дату начала наблюдений, чтобы при вычислении разности между днями массив остался той же длины
gap_days = gap_days.map(lambda x: [x[0]] + x).apply(np.diff).map(lambda x: [xi.days for xi in x])

#Превращаем все в серию, отсортированную по loc_id
gap_days = pd.Series(np.concatenate(gap_days.values))

In [15]:
df['delta_days_of_records'] = gap_days
df['days_without_rain'] = 0

In [16]:
for i in range(1,df.shape[0]):
    if (df.loc[i,'delta_days_of_records'] > 2) or (df.loc[i, 'was_NA_RainToday'] == 1):#((df.loc[i,'RainToday'] != 0) and (df.loc[i,'RainToday'] != 2)):
        df.loc[i, 'days_without_rain'] = -10;
    else:
        if df.loc[i,'RainToday'] == 0: #Если дождь
            df.loc[i, 'days_without_rain'] = 0
        else:
            if df.loc[i, 'loc_id'] == df.loc[i-1, 'loc_id']:
                df.loc[i, 'days_without_rain'] = df.loc[i-1, 'days_without_rain'] + 1

In [17]:
df.head(10)

Date  loc_id   MinTemp   MaxTemp  Rainfall  Evaporation  Sunshine  \
0 2008-07-02     0.0  0.531328  0.381226  0.002176     0.009655  0.545455   
1 2008-07-03     0.0  0.368421  0.367816  0.000000     0.012414  0.146853   
2 2008-07-07     0.0  0.403509  0.293103  0.044070     0.031724  0.076923   
3 2008-07-08     0.0  0.345865  0.337165  0.046246     0.004138  0.146853   
4 2008-07-09     0.0  0.423559  0.352490  0.004897     0.011034  0.055944   
5 2008-07-10     0.0  0.451128  0.329502  0.024483     0.008276  0.503497   
6 2008-07-13     0.0  0.448622  0.400383  0.006801     0.017931  0.636364   
7 2008-07-16     0.0  0.370927  0.348659  0.000544     0.011034  0.265734   
8 2008-07-17     0.0  0.458647  0.360153  0.000544     0.011034  0.244755   
9 2008-07-20     0.0  0.461153  0.308429  0.023123     0.005517  0.020979   

   WindGustDir  WindGustSpeed  WindDir9am  ...  was_NA_Cloud3pm  \
0     0.800000       0.224806    0.733333  ...              1.0   
1     0.866667       0.108527    0.333333  ...              1.0   
2     1.000000       0.310078    0.933333  ...              1.0   
3     0.800000       0.286822    0.800000  ...              1.0   
4     0.466667       0.271318    0.200000  ...              1.0   
5     1.000000       0.356589    1.000000  ...              1.0   
6     0.400000       0.372093    0.200000  ...              1.0   
7     0.400000       0.255814    0.333333  ...              1.0   
8     0.466667       0.294574    0.200000  ...              1.0   
9     0.200000       0.310078    0.200000  ...              1.0   

   was_NA_Temp9am  was_NA_Temp3pm  was_NA_RainToday  year     month  \
0             0.0             0.0               0.0   0.1  0.545455   
1             0.0             0.0               0.0   0.1  0.545455   
2             0.0             0.0               0.0   0.1  0.545455   
3             0.0             0.0               0.0   0.1  0.545455   
4             0.0             0.0               0.0   0.1  0.545455   
5             0.0             0.0               0.0   0.1  0.545455   
6             0.0             0.0               1.0   0.1  0.545455   
7             0.0             0.0               0.0   0.1  0.545455   
8             0.0             0.0               0.0   0.1  0.545455   
9             0.0             0.0               1.0   0.1  0.545455   

   day_of_week  week_of_year  delta_days_of_records  days_without_rain  
0     0.333333      0.500000                      0                  0  
1     0.500000      0.500000                      1                  1  
2     0.000000      0.519231                      4                -10  
3     0.166667      0.519231                      1                  0  
4     0.333333      0.519231                      1                  0  
5     0.500000      0.519231                      1                  0  
6     1.000000      0.519231                      3                -10  
7     0.333333      0.538462                      3                -10  
8     0.500000      0.538462                      1                 -9  
9     1.000000      0.538462                      3                -10  

[10 rows x 52 columns]

# FIT -> PREDICT

In [18]:
X = df.drop(['RainTomorrow', 'Date', 'delta_days_of_records'], axis=1)
y = df['RainTomorrow']

skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
model = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)

scores = cross_val_score(model, X, y, cv=skf, scoring='accuracy')
print(f"Результаты: {scores.mean():.4f} ± {scores.std():.4f}")

Результаты: 0.8820 ± 0.0021


mean Результаты: 0.8582 ± 0.0016\
median Результаты: 0.8584 ± 0.0020\
2 neighbor KNN Результаты: 0.8820 ± 0.0021

In [19]:
print(scores)

[0.87893706 0.88406088 0.88014266 0.88380972 0.88290551]


# Feature importance

In [20]:
model.fit(X, y)
importance_df = pd.DataFrame({'importance' : model.feature_importances_, 'columns' : X.columns}).sort_values('importance', ascending=False)
importance_df

importance               columns
5     0.124334              Sunshine
13    0.118659           Humidity3pm
17    0.057419              Cloud3pm
15    0.054133           Pressure3pm
14    0.045233           Pressure9am
7     0.044177         WindGustSpeed
3     0.038166              Rainfall
12    0.036156           Humidity9am
19    0.032374               Temp3pm
1     0.028598               MinTemp
16    0.027938              Cloud9am
4     0.027733           Evaporation
2     0.027501               MaxTemp
18    0.026852               Temp9am
22    0.025088             Longitude
47    0.022304          week_of_year
11    0.021263          WindSpeed3pm
21    0.019902              Latitude
10    0.019274          WindSpeed9am
8     0.019100            WindDir9am
6     0.018827           WindGustDir
9     0.018798            WindDir3pm
20    0.018445             RainToday
0     0.017621                loc_id
23    0.017601             Elevation
44    0.015572                  year
48    0.015025     days_without_rain
45    0.013993                 month
46    0.013895           day_of_week
28    0.004524       was_NA_Sunshine
40    0.004404       was_NA_Cloud3pm
27    0.004245    was_NA_Evaporation
39    0.004124       was_NA_Cloud9am
37    0.001893    was_NA_Pressure9am
26    0.001891       was_NA_Rainfall
43    0.001860      was_NA_RainToday
38    0.001837    was_NA_Pressure3pm
29    0.001710    was_NA_WindGustDir
30    0.001699  was_NA_WindGustSpeed
31    0.001426     was_NA_WindDir9am
36    0.000810    was_NA_Humidity3pm
32    0.000733     was_NA_WindDir3pm
42    0.000595        was_NA_Temp3pm
35    0.000556    was_NA_Humidity9am
34    0.000544   was_NA_WindSpeed3pm
33    0.000395   was_NA_WindSpeed9am
41    0.000350        was_NA_Temp9am
24    0.000246        was_NA_MinTemp
25    0.000177        was_NA_MaxTemp

# Использование OneHotEncoder

In [21]:
df = pd.read_csv('datasets/train_df.csv')
coordinate_df = pd.DataFrame(coordinates)

In [22]:
#Соединяем два датафрейма
df = pd.merge(df, coordinate_df, how='outer', on='Location')

In [23]:
#df['RainTomorrow'].plot(kind='hist')
df['Date'] = pd.to_datetime(df['Date'], format='%Y-%m-%d')

In [24]:
category_list = ['Location', 'RainToday', 'RainTomorrow', 'WindGustDir', 'WindDir9am', 'WindDir3pm']

In [25]:
#Создаем отдельный столбец для каждой фичи, в значениях которой встречается Nan
for column in df.columns:
    if df[column].isna().sum() > 0:
        df['was_NA_' + column] = 0
        df.loc[df[column].isna() == True, 'was_NA_' + column] = 1

#Категориальные заполняем Unknown, а числовые - медианой
df = df.apply(lambda x: x.fillna('Unknown') if x.name in category_list else x.fillna(x.median()))

In [26]:
y = df['RainTomorrow']
df = pd.get_dummies(df, columns=category_list)
df = df.drop(columns=['RainTomorrow_No', 'RainTomorrow_Yes'])
df

Date  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0     2015-04-08      9.9     19.9       0.0          2.2       8.5   
1     2016-04-28     16.4     26.9       0.0          4.8       8.5   
2     2010-06-13      4.3     13.4       0.0          4.8       1.1   
3     2013-12-12     13.8     22.6       0.0          4.4       8.8   
4     2013-01-23     15.5     30.6       0.0          7.0      13.2   
...          ...      ...      ...       ...          ...       ...   
99530 2016-10-04      7.3     23.2       0.0          7.4       8.5   
99531 2016-03-05     25.1     40.1       0.0         14.8      11.3   
99532 2011-08-04     13.3     23.6       0.0         11.4       8.2   
99533 2010-09-14      9.2     18.0       5.2          3.0       6.3   
99534 2015-05-31      5.6     17.3       0.0          4.8       8.5   

       WindGustSpeed  WindSpeed9am  WindSpeed3pm  Humidity9am  ...  \
0               31.0           9.0          15.0         61.0  ...   
1               28.0           4.0           7.0         84.0  ...   
2               20.0           0.0          13.0         74.0  ...   
3               33.0          19.0          15.0         72.0  ...   
4               35.0           9.0           9.0         46.0  ...   
...              ...           ...           ...          ...  ...   
99530           61.0          31.0          37.0         49.0  ...   
99531           50.0          20.0          28.0         36.0  ...   
99532           48.0          19.0          19.0         67.0  ...   
99533           33.0          24.0          19.0         74.0  ...   
99534           41.0           9.0          24.0         88.0  ...   

       WindDir3pm_NW  WindDir3pm_S  WindDir3pm_SE  WindDir3pm_SSE  \
0              False         False          False           False   
1              False         False          False           False   
2              False         False          False           False   
3              False         False          False           False   
4              False         False          False           False   
...              ...           ...            ...             ...   
99530          False         False          False           False   
99531          False         False          False           False   
99532          False         False          False           False   
99533          False          True          False           False   
99534          False         False          False           False   

       WindDir3pm_SSW  WindDir3pm_SW  WindDir3pm_Unknown  WindDir3pm_W  \
0               False           True               False         False   
1               False          False               False         False   
2               False           True               False         False   
3               False           True               False         False   
4               False           True               False         False   
...               ...            ...                 ...           ...   
99530           False          False               False         False   
99531           False          False               False         False   
99532           False          False               False         False   
99533           False          False               False         False   
99534           False           True               False         False   

       WindDir3pm_WNW  WindDir3pm_WSW  
0               False           False  
1               False           False  
2               False           False  
3               False           False  
4               False           False  
...               ...             ...  
99530           False            True  
99531           False           False  
99532            True           False  
99533           False           False  
99534           False           False  

[99535 rows x 143 columns]

In [27]:
df['year'] = df['Date'].apply(lambda x: x.year)
df['month'] = df['Date'].apply(lambda x: x.month)
df['day_of_week'] = df['Date'].apply(lambda x: x.dayofweek)
df['week_of_year'] = df['Date'].apply(lambda x: x.weekofyear)
#df = df.sort_values(['loc_id', 'Date']).reset_index().drop('index', axis=1)
df.head(10)

Date  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0 2015-04-08      9.9     19.9       0.0          2.2       8.5   
1 2016-04-28     16.4     26.9       0.0          4.8       8.5   
2 2010-06-13      4.3     13.4       0.0          4.8       1.1   
3 2013-12-12     13.8     22.6       0.0          4.4       8.8   
4 2013-01-23     15.5     30.6       0.0          7.0      13.2   
5 2008-09-26     14.0     28.0       0.0          5.6      11.5   
6 2014-07-30     12.7     17.3       3.2          0.4       8.5   
7 2011-05-26      9.7     15.5       0.6          0.6       3.2   
8 2015-09-19      5.8     21.1       0.0          4.8       8.5   
9 2017-03-02     21.1     36.2       0.0          4.8       8.5   

   WindGustSpeed  WindSpeed9am  WindSpeed3pm  Humidity9am  ...  \
0           31.0           9.0          15.0         61.0  ...   
1           28.0           4.0           7.0         84.0  ...   
2           20.0           0.0          13.0         74.0  ...   
3           33.0          19.0          15.0         72.0  ...   
4           35.0           9.0           9.0         46.0  ...   
5           39.0          15.0          17.0         18.0  ...   
6           43.0          15.0          20.0         88.0  ...   
7           28.0           4.0           7.0         76.0  ...   
8           24.0           7.0           9.0         60.0  ...   
9           33.0           4.0          11.0         46.0  ...   

   WindDir3pm_SSW  WindDir3pm_SW  WindDir3pm_Unknown  WindDir3pm_W  \
0           False           True               False         False   
1           False          False               False         False   
2           False           True               False         False   
3           False           True               False         False   
4           False           True               False         False   
5           False          False               False         False   
6           False          False               False         False   
7           False           True               False         False   
8           False          False               False          True   
9            True          False               False         False   

   WindDir3pm_WNW  WindDir3pm_WSW  year  month  day_of_week  week_of_year  
0           False           False  2015      4            2            15  
1           False           False  2016      4            3            17  
2           False           False  2010      6            6            23  
3           False           False  2013     12            3            50  
4           False           False  2013      1            2             4  
5           False           False  2008      9            4            39  
6           False           False  2014      7            2            31  
7           False           False  2011      5            3            21  
8           False           False  2015      9            5            38  
9           False           False  2017      3            3             9  

[10 rows x 147 columns]

In [28]:
X = df.drop(['Date'], axis=1)
#y = df['RainTomorrow']

skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
model = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)

scores = cross_val_score(model, X, y, cv=skf, scoring='accuracy')
print(f"Результаты: {scores.mean():.4f} ± {scores.std():.4f}")

Результаты: 0.8571 ± 0.0024


# Результаты сравнения

Label Encoding Результаты: 0.8583 ± 0.0020\
One Hot Encoding Результаты: 0.8571 ± 0.0024